common/thorhelper/thormeta.cpp

```diff
@@ -0,0 +1,531 @@
+    if (numParts > maxParts)
+        maxParts = numParts;
+}
+
+
+void CLogicalFileCollection::calcLocations()
+{
+    //If there is only a single file, then the global part numbers and base offsets match the values within the file.
+    if (files.ordinality() == 1)
+        return;
+
+    //If there are multiple parts then they need to be calculated within the interleaved files in the superfile.
+    unsigned superPartNum = 0;
+    offset_t baseOffset = 0;
+    for (unsigned part=0; part < maxParts; part++)
+    {
+        ForEachItemIn(i, files)
+        {
+            CLogicalFile & cur = files.item(i);
+            unsigned numParts = cur.getNumParts();
+            if (numParts > part)
+            {
+                cur.noteLocation(part, superPartNum, baseOffset);
+                superPartNum++;
+                baseOffset += cur.getPartSize(part);
+            }
+        }
+    }
+    //We may need a function to map from super part numbers to a file/part - in which case we'll need to create a reverse
+    //mapping {fileIndex,part} if it needs to be done efficiently
+}
+
+/*
+Calculate which parts of which files will be included in this channel
+
+There is an implicit ordering of the file parts - file1, part1, file2, part1, file<n> part1, file1, part2 etc.
+
+if distribution and ordering are preservered:
+    files are processed in order, multiple parts can be processed on a single node as long as global ordering is retained.
+
+if ordering is preserved and distribution does not need to be:
+    files must be processed in order, parts can be processed on any node as long as global ordering is retained.
+    e.g. if number of reading nodes is > number of file parts, then part1s can be split between nodes1 and 2, part2 between nodes 3 and 4
+         or if multiple files, node1 can process file1 part1, node2 file2 part1, node3 file1 part2, node4 file1 part2
+
+if distribution is preserved, but ordering does not need to be:
+    on bare metal, it might be better to have part1,part<n+1> on node1, part2,part<n+2> on node2 since first parts more likely to be local
+
+if neither distribution or ordering are preserved
+    if non local storage, could try and ensure that workload is even across the nodes.  Could split parts between nodes (but beware
+    of making the read parts too small e.g. for blob read granularity)
+
+on bare metal, if channels is a multiple of the maximum number of parts it is likely to be worth reading the same as inorder
+*/
+
+void CLogicalFileCollection::calcPartition(SliceVector & slices, unsigned numChannels, unsigned channel, bool preserveDistribution, bool preserveOrder)
+{
+    calcLocations();
+
+    //MORE: Revisit and improve the following code to optimize cases detailed above, once they are likely to occur.
+    //Likely to require code generator/language improvements first
+    unsigned partsPerNode = (maxParts < numChannels) ? 1 : (maxParts + numChannels - 1) / numChannels;
```

Comment:
Yes, that is the case.  However with the current language semantics that redistribution is not possible, and until the semantics of SORT are extended it will be very hard to spot any cases automatically.  You could explicitly indicate the file can be redistributed, but that it currently unlikely to happen.
In practice we tend to stick to well defined numbers of parts that are multiples of each other.  The common example is a single part file being read <N> ways.